Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,577 @@
+                UInt16(uint8Tuple: (self.address[idx], self.address[idx + 1]))
+            })
+            
+            var bucketStrings: Array<String> = bucketValues.map {String(uint16Hex: $0)}
+            
+            // find most consecutive zeros and remove them
+            var consecutiveZeros = [0, 0, 0, 0, 0, 0, 0, 0]
+            var consecutiveCounter = 0
+            for (idx, value) in bucketValues.enumerated() {
+                switch value {
+                case 0: consecutiveCounter += 1
+                default: consecutiveCounter = 0
+                }
+                consecutiveZeros[idx] = consecutiveCounter
+            }
+            let mostConsecutiveZeros = consecutiveZeros.max()
+            
+            if let mostConsecutiveZeros = mostConsecutiveZeros {
+                if mostConsecutiveZeros > 0 {
+                    let consecutiveZeroIndex = consecutiveZeros.firstIndex(of: mostConsecutiveZeros)
+                    
+                    if let endCompression = consecutiveZeroIndex {
+                        let startCompression = endCompression - mostConsecutiveZeros + 1
+                        
+                        switch (startCompression, endCompression) {
+                        case (0, 7): bucketStrings.replaceSubrange(startCompression...endCompression, with: ["::"])
+                        case (0, _): bucketStrings.replaceSubrange(startCompression...endCompression, with: [":"])
+                        case (_, 7): bucketStrings.replaceSubrange(startCompression...endCompression, with: [":"])
+                        case (_, _): bucketStrings.replaceSubrange(startCompression...endCompression, with: [""])
+                        }
+                    }
+                }
+            }
+            
+            return bucketStrings.joined(separator: ":")
+        }
+        
+        /// A human-readable description of this `IPv6Address`. Mostly useful for logging.
+        public var description: String {
+            return "[IPv6]\(self.ipAddressString)"
+        }
+        
+        /// Get the libc address for an IPv6 address.
+        public var posix: in6_addr {
+            get {
+                return in6_addr.init(__u6_addr: .init(__u6_addr8: self.address.bytes))
+            }
+        }
+        
+        /// Creates a new `IPv6Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv6 address.
+        public init(address: IPv6Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv6Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        public init<Bytes: Collection>(packedBytes bytes: Bytes) where Bytes.Element == UInt8 {
+            var ipv6Bytes = IPv6Bytes((0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0))
+            
+            for (idx, elt) in bytes.prefix(16).enumerated() {
```

Comment:
Rather than use a `prefix` here we should `throw` if the collection does not contain exactly 16 elements.